js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,196 @@
+import Model from '../Model';
+
+export interface Page<TModel> {
+  number: number;
+  items: TModel[];
+}
+
+export interface PaginationLocation {
+  page: number;
+  startIndex?: number;
+  endIndex?: number;
+}
+
+export default abstract class PaginatedListState<T extends Model> {
+  protected location!: PaginationLocation;
+  protected offset: number;
+
+  protected pages: Page<T>[] = [];
+  protected params: any = {};
+
+  protected initialLoading: boolean = false;
+  protected loadingPrev: boolean = false;
+  protected loadingNext: boolean = false;
+
+  hasPrev: boolean = false;
+  hasNext: boolean = false;
+
+  protected constructor(offset: number = 20) {
+    this.offset = offset;
+  }
+
+  abstract get type(): string;
+
+  protected parseResults(pageNum, results: T[]) {
+    const page = {
+      number: pageNum,
+      items: results,
+    };
+
+    if (this.isEmpty() || pageNum > this.getNextPageNumber() - 1) {
+      this.pages.push(page);
+    } else {
+      this.pages.unshift(page);
+    }
+
+    this.hasNext = !!results.payload?.links?.next;
+    this.hasPrev = !!results.payload?.links?.prev;
+
+    m.redraw();
+  }
+
+  /**
+   * Load a new page of results.
+   */
+  protected loadPage(page = 1): Promise<T[]> {
+    const params = this.requestParams();
+    params.page = { offset: this.offset * (page - 1) };
+
+    if (Array.isArray(params.include)) {
+      params.include = params.include.join(',');
+    }
+
+    return app.store.find(this.type, params);
+  }
+
+  public hasItems(): boolean {
+    return !!this.getAllItems().length;
+  }
+
+  public isEmpty(): boolean {
+    return !this.isInitialLoading() && !this.hasItems();
+  }
+
+  public getLocation(): PaginationLocation {
+    return this.location;
+  }
+
+  protected getAllItems(): T[] {
+    return this.pages.map((pg) => pg.items).flat();
+  }
+
+  public isInitialLoading(): boolean {
+    return this.initialLoading;
+  }
+  public isLoadingPrev(): boolean {
+    return this.loadingPrev;
+  }
+  public isLoadingNext(): boolean {
+    return this.loadingNext;
+  }
+
+  /**
+   * Get the parameters that should be passed in the API request.
+   * Do not include page offset unless subclass overrides loadPage.
+   *
+   * @abstract
+   * @see loadPage
+   */
+  protected requestParams(): any {
+    return {};
```

Comment:
This should probably use `params`.

Also, would like to see the qBuilder included into this system

Reply:
You mean return `this.params` by default?

Reply:
Implemented by Ian and myself in https://github.com/FriendsOfFlarum/user-directory/commit/bc4ffe801c1d4feac5f01aa4b05fbcd787a65e91

Essentially, if we have 2 different places in extensions that are trying to modify the `q` parameter, it gets messy, fast. Instead, we use a qBuilder object. Keys are just unique identifiers, and the values are space-joined to obtain the actual `q` value.

Reply:
Hm, I feel like we probably want a separate PR for that functionality. We probs want to discuss it a bit more outside this PR.